content/posts/redis-assisted-client-side-caching-in-python.md

```diff
@@ -0,0 +1,213 @@
+
+```python
+    def _handler(self, message):
+        ''' Handles invalidation messages '''
+        slot = message['data']
+        self.invalidate(slot)
+
+    def invalidate(self, slot):
+        ''' Invalidates a slot's keys '''
+        slot = int(slot)
+        while self.slots[slot]:
+           key = self.slots[slot].pop()
+           del self.cache[key]
+```
+
+Invalidation is just a matter of popping keys, one by one, from the respective slot's set and deleting them from the cache. Lastly, the manager exposes a factory method, `get_connection()` that's used by the code for getting new cached connections:
+
+```python
+    def get_connection(self, *args, **kwargs):
+        ''' Returns a cached Redis connection '''
+        conn = CachedRedis(self, connection_pool=self.pool, *args, **kwargs)
+        return conn
+```
+
+## Some measurements
+This post isn't about benchmarking or Python's performance per se, but it’s important to understand the mechanism's impact. For that purpose, I've used the [benchmark.py](https://github.com/itamarhaber/rsacsc-py/blob/master/benchmark.py) script on a 2019 MacBook Pro with a Redis instance running locally using defaults (except that I turned off snapshotting).
+
+Before performing the tests, the benchmark script populates the database with 1.000 keys and sets up the cache manager with a capacity for 100. It then runs several timed tests to measure performance. Each test is repeated five times for both types of connections: regular and cached.
+
+The result of the first test actually demonstrates one of caching's _disadvantages_: cache misses. In this test, `single_read`, we read every key from the entire database just once, so every access to the local cache results in a miss:
+
+<div id="chart_1"></div>
+
+Note that the averages are computed only for the last runs, so every first run in the series is considered a warmup. The averages above show that the missed cache reads add almost 13ms for every 1,000 reads, roughly an 18% latency increase.
+
+However, repeating the test on a dataset that fits into the cache—that is, only 100 keys—shows more encouraging results. While the first cached run shows an increase in latency, subsequent ones reduce latency by two orders of magnitude:
+
+<div id="chart_2"></div>
+
+The next test is named `eleven_reads` because it reads every key in the database once along with 10 other keys that are always the same ones. This highly synthetic use case provides an even more dramatic proof of the cache's benefit (despite that not being the purpose, per se).
+
+<div id="chart_3"></div>
+
+The last test extends `eleven_reads` with an additional write request to one of the 10 constant keys, which triggers the invalidation of a part of the cache. Latency of cached runs increase slightly, both because of the extra write command but also due to the need to refetch the contents of the cache:
```

Comment:
"Latency of cached runs increase slightly" => "...increases slightly"

Reply:
👍